src/Workspaces/CSharp/Portable/VirtualChars/CSharpVirtualCharService.cs

```diff
@@ -0,0 +1,255 @@
+            {
+                Debug.Assert(false, "This should not be reachable as long as the compiler added no diagnostics.");
+                return default;
+            }
+
+            var startIndexInclusive = StartDelimeter.Length;
+            var endIndexExclusive = tokenText.Length - EndDelimeter.Length;
+
+            var result = ArrayBuilder<VirtualChar>.GetInstance();
+            try
+            {
+                var offset = token.SpanStart;
+                for (var index = startIndexInclusive; index < endIndexExclusive;)
+                {
+                    if (tokenText[index] == '\\')
+                    {
+                        if (!TryAddEscape(result, tokenText, offset, index))
+                        {
+                            return default;
+                        }
+
+                        index += result.Last().Span.Length;
+                    }
+                    else
+                    {
+                        result.Add(new VirtualChar(tokenText[index], new TextSpan(offset + index, 1)));
+                        index++;
+                    }
+                }
+
+                return result.ToImmutable();
+            }
+            finally
+            {
+                result.Free();
+            }
+        }
+
+        private bool TryAddEscape(
+            ArrayBuilder<VirtualChar> result, string tokenText, int offset, int index)
+        {
+            // Copied from Lexer.ScanEscapeSequence.
+            Debug.Assert(tokenText[index] == '\\');
+
+            var ch = tokenText[index + 1];
+            switch (ch)
+            {
+            // escaped characters that translate to themselves
+            case '\'':
+            case '"':
+            case '\\':
+                break;
+            // translate escapes as per C# spec 2.4.4.4
+            case '0':
+                ch = (char)0;
+                break;
+            case 'a':
+                ch = '\u0007';
+                break;
+            case 'b':
+                ch = '\u0008';
+                break;
+            case 'f':
+                ch = '\u000c';
+                break;
+            case 'n':
+                ch = '\u000a';
+                break;
+            case 'r':
+                ch = '\u000d';
+                break;
+            case 't':
+                ch = '\u0009';
+                break;
+            case 'v':
+                ch = '\u000b';
+                break;
+            case 'x':
+            case 'u':
+            case 'U':
+                return TryAddComplexEscape(result, tokenText, offset, index, ch);
+            default:
+                Debug.Assert(false, "This should not be reachable as long as the compiler added no diagnostics.");
+                return false;
+            }
+
+            result.Add(new VirtualChar(ch, new TextSpan(offset + index, 2)));
+            return true;
+        }
+
+        private bool TryAddComplexEscape(
```

Comment:
Add a `<summary>` giving a quick overview of what a "complex escape" is.

Reply:
Sure.  I can doc that.

Reply:
Changed the name to be something liek TryAddSingleCharEscape and TryAddMultiCharEscape.